### PR TITLE
Bulk pool creation

### PR DESCRIPTION
**Pull request by apmasell**

### CODE REVIEW

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditPoolController.java [hunk 1]
```diff
@@ -267,4 +270,27 @@ public void addSequencingParameters(ModelMap model) throws IOException {
     array.addAll(sequencingParameters);
     model.put("sequencingParametersJson", array.toString());
   }
+
+  private final BulkEditTableBackend<Pool, PoolDto> dilutionBulkEditBackend = new BulkEditTableBackend<Pool, PoolDto>(
```
Comment (callunity) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditPoolController.java]: `LibraryDilution` for the generics?

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/PoolRestController.java [hunk 2]
```diff
@@ -105,13 +105,32 @@ public void setDilutionService(LibraryDilutionService dilutionService) {
...
-    return mapper.writeValueAsString(p);
+    return Dtos.asDto(p, true);
+  }
+
+  @RequestMapping(method = RequestMethod.POST, produces = "application/json")
+  @ResponseBody
+  public PoolDto createPool(@RequestBody PoolDto pool, UriComponentsBuilder uriBuilder, HttpServletResponse response)
+      throws IOException {
+    Long id = poolService.savePool(Dtos.to(pool));
+    UriComponents uriComponents = uriBuilder.path("/{id}").buildAndExpand(id);
```
Comment (callunity) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/PoolRestController.java]: `"/pool/{id}"`?

File: miso-web/src/main/webapp/scripts/hot_dilution.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                .maybeGetProperty(
+                    Utils.array
+                        .findFirstOrNull(
+                            function(tarSeq) {
+                              return flat.targetedSequencingAlias == tarSeq.alias && tarSeq.kitDescriptors
+                                  .indexOf(dil.library.kitDescriptorId) != -1;
+                            }, Constants.targetedSequencings), 'id');
+          },
+          depends : '*start', // This is a dummy value that gets this run on
+          // creation only
```
Comment (callunity) [inline miso-web/src/main/webapp/scripts/hot_dilution.js]: Dilution creation (as opposed to update) or table creation?

File: miso-web/src/main/webapp/scripts/hot_dilution.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+      },
+      {
+        name : 'Create one pool',
+        action : function(ids) {
+          window.location = window.location.origin + '/miso/library/dilution/bulk/merge/' + ids
+              .join(',');
+        }
+      },
+      {
+        name : 'Create many pools',
```
Comment (callunity) [inline miso-web/src/main/webapp/scripts/hot_dilution.js]: This is somewhat ambiguous. `Create one pool per dilution`?

File: miso-web/src/main/webapp/scripts/hot_dilution.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  bulkActions : [
+      {
+        name : 'Edit',
+        action : function(ids) {
+          window.location = window.location.origin + '/miso/library/dilution/bulk/edit/' + ids
+              .join(',');
+        }
+      },
+      {
+        name : 'Create one pool',
```
Comment (callunity) [inline miso-web/src/main/webapp/scripts/hot_dilution.js]: One pool to rule them all...

File: miso-web/src/main/webapp/scripts/hot_pool.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    callback(config)
+  },
+  fixUp : function(lib, errorHandler) {
+  },
+  createColumns : function(config, create, data) {
+    return [
+        {
+          header : 'Pool Alias',
+          data : 'alias',
+          readOnly : true,
```
Comment (callunity) [inline miso-web/src/main/webapp/scripts/hot_pool.js]: Alias should be user-editable.

File: miso-web/src/main/webapp/scripts/hot_dilution.js (hunk 5 above)
Comment (apmasell) [inline miso-web/src/main/webapp/scripts/hot_dilution.js]: I don't know what to call this one (or the next one). “Create one pool per dilution” is really long for a button.

**Comment by tonydebat** [comment]
Error when making dilutions from libraries. (Making pools from dilutions works without error.)

```
2017-06-20 13:55:07 ERROR: 500 error handling REST request
java.lang.NullPointerException
	at uk.ac.bbsrc.tgac.miso.core.data.AbstractLibrary.setPaired(AbstractLibrary.java:312)
	at uk.ac.bbsrc.tgac.miso.dto.Dtos.to(Dtos.java:1108)
	at uk.ac.bbsrc.tgac.miso.dto.Dtos.to(Dtos.java:1285)
	at uk.ac.bbsrc.tgac.miso.webapp.controller.rest.LibraryDilutionRestController.createDilution(LibraryDilutionRestController.java:128)
	at sun.reflect.GeneratedMethodAccessor359.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:133)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:116)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:827)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:738)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:963)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:897)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:872)
```

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/PoolRestController.java (hunk 2 above)
Comment (apmasell) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/PoolRestController.java]: I believe it's relative to the current path. Now that we return the DTO, I think we can not bother.

**Comment by apmasell** [comment]
The NPE has been fixed.

**Comment by callunity** [comment]
What do you think about setting "Ready to Run" to `True` by default, instead of the current `Unknown`? Also, since it's a `boolean` field, shouldn't the options only be `True`/`False`?

**Comment by callunity** [comment]
I have mild concerns about the default pooled-together pool alias being `PROJ_POOL` because it's so easy to make many of those with the same alias and then you have to click into the pool to distinguish between them, but we can leave that be and see if the lab complains.

**Pull request merged**